src/storage/src/source/persist_source.rs

```diff
@@ -131,35 +131,48 @@ where
-            // as a frontier in `Progress` messages while a timely source usually only has a single
-            // `Capability` that it can downgrade. We can work around that by using a
-            // `CapabilitySet`, which allows downgrading the contained capabilities to a frontier
-            // but `util::source` gives us both a vanilla `Capability` and a `CapabilitySet` that
-            // we have to keep downgrading because we cannot simply drop the one that we don't
-            // need.
-            let mut current_ts = 0;
-
-            move |cap, output| {
+            move |cap_set, output| {
                 let mut context = Context::from_waker(&waker);
                 // Bound execution of operator to prevent a single operator from hogging
                 // the CPU if there are many messages to process
                 let timer = Instant::now();
 
                 while let Poll::Ready(item) = pinned_stream.as_mut().poll_next(&mut context) {
                     match item {
-                        Some(Ok(ListenEvent::Progress(upper))) => match upper.into_option() {
-                            Some(ts) => {
-                                current_ts = ts;
-                                cap.downgrade(&ts);
+                        Some(Ok(ListenEvent::Progress(upper))) => {
+                            cap_set.downgrade(upper.iter());
+
+                            if upper.is_empty() {
+                                return SourceStatus::Done;
                             }
-                            None => return SourceStatus::Done,
-                        },
+                        }
                         Some(Ok(ListenEvent::Updates(mut updates))) => {
                             // This operator guarantees that its output has been advanced by `as_of.
                             // The persist SnapshotIter already has this contract, so nothing to do
                             // here.
-                            let cap = cap.delayed(&current_ts);
+
+                            if updates.is_empty() {
+                                continue;
+                            }
+
+                            // Determine the minimum timestamp and emit all
+                            // updates at that timestamp. We could be more
+                            // fine-grained and first partition the updates by
+                            // timestamp. But then we'd have multiple give_vec()
+                            // calls below.
+                            // TODO(aljoscha): I don't like having to iterate
+                            // through all updates in order to determine a
+                            // timestamp. Maybe persist `listen()` should emit a
+                            // timestamp for the batch, similar to how inputs
+                            // work in
+                            // timely.
+                            let ts = updates
```

Comment:
I didn't understand this part, why can't we emit each update at its timestamp? Is this trying to reduce the number of unique timely timestamps?

Reply:
We need to get a capability  from the cap set and open a session with it. And then I give all the updates in a single `give_vec` call. If I want to emit each update at a timely timestamp that corresponds to its timestamp I'd have to partition the updates by timestamp and then do that logic per timestamp (it's above this part in the comment). You think I should do that here? I think in practice there won't be many different timestamps in a batch, I think it's likely that there is mostly only one.

Reply:
> If I want to emit each update at a timely timestamp that corresponds to its timestamp I'd have to partition the updates by timestamp and then do that logic per timestamp

Do we have to partition? I thought it would be something like this:

```rust
for (update, ts, diff) in updates {
    let cap = cap_set.delayed(ts);
    output.session(&cap).give((update, ts, diff));
}
````

The logic of selecting the minimum timestamp only works in totally ordered time so up to you whether you want to leave that in here. I think the fully fledged logic for what you're trying to do is this one https://github.com/TimelyDataflow/differential-dataflow/blob/master/src/operators/arrange/arrangement.rs#L576-L661

Reply:
The loop above is a bit of a timely anti-pattern; each call to `session` with a different time will flush the internals and close off a timely internal message batch. If you were able to sort `updates` by `ts` first, and there aren't many of them, it wouldn't be as bad.

If you want to minimize the number of timely batches, you grab each capability in the set in turn, peel out the  elements of `updates` greater or equal to that capability, and send those. By the end you should have sent everything, so if anything is left there is some sort of bug.

Reply:
> If you want to minimize the number of timely batches, you grab each capability in the set in turn, peel out the elements of updates greater or equal to that capability, and send those. By the end you should have sent everything, so if anything is left there is some sort of bug.

I think I'll do that, thanks! Currently, this won't make a difference from my solution because we have totally ordered times and there is at most one capability. But that will make it work with partially ordered times, should we ever have them here.